wordpress-mobile/WordPress-Android
# Update free domain suggestions UI

Resolves #17899 

This PR:
- ⊕ Adds the new UI for domain suggestions using Compose #17899
- ⊕ Refactors view logic for domain suggestions:
  - ⊕ Migrates the domain suggestion view-holder to view binding
  - ⊕ Refactors the UI state of domain list items to separate the old vs new one
  - ⊖ Removes the repetitive `Domain` prefixes from ui state class names when it can be deduced
  - ⊜ Converts single line vm methods from block to expression body

> **Note**
> To easily review the many changes please go commit by commit.

## To Test
#### Prerequisites:

<details>
<summary>◐ Toggle the Site Creation Domain Purchasing feature (<code>SiteCreationDomainPurchasingFeatureConfig</code>)</summary>

1. Go to `Me` → `App Settings` → `Debug settings`
3. Scroll to the `Features in development` section
4. Tap on `SiteCreationDomainPurchasingFeatureConfig`
6. Tap `RESTART THE APP` button
---
</details>

### 🅰️ Control Variation

1. Open the Jetpack app and login
2. ◐ Toggle the `Site Creation Domain Purchasing` feature flag `OFF` (should be the case)
3. Start the Site Creation flow: `Open Site Picker` → `➕` → `Create WordPress.com site`
4. Continue the flow till `Choose a domain` screen
5. **Verify** the expected **results** functionality is maintained by the refactoring:
   1. Enter `test` in the search field
   2. **Expect** the first result to be the unavailable `test.wordpress.com` followed by available one(s)
   3. Tap on the `✗` button to clear the search query
   4. **Expect** The results to be replaced by the placeholder illustration
   5. Enter `tested` in the search field
   6. **Expect** all results to be available and selectable
6. **Verify** the expected **error** functionality keeps working:</summary>
   1. Tap on the `✗` button to clear the search query
   2. Turn off internet connectivity on your testing device
   3. Enter `testet` in the search field
   4. **Expect** The result to be `There is no network available`
   5. Turn on internet connectivity
   6. Tap the error item (e.g. on `Retry`)
   7. **Expect** the results to be updated
   8. Replace the `t` in the search field with `%`
   9. **Expect** the error message for invalid characters

### 🅱️ Treatment Variation

1. ◐ Toggle the `Site Creation Domain Purchasing` feature flag `ON`
7. Start the Site Creation flow: `Open Site Picker` → `➕` → `Create WordPress.com site`
8. Continue the flow till `Choose a domain` screen
9. **Verify** free and paid domains are suggested
10. Tap on a free domain
11. **Verify** the `Create Site` button is **displayed**
12. Tap on a paid domain
13. **Verify** the `Create Site` button is **hidden**

### Screenshots

| 🅰️ Control | 🅱️ Treatment |
| - | - |
| Before | After |
| <img width="300" src="https://user-images.githubusercontent.com/4588074/220423925-9ff34be3-3680-40e2-825b-3a94a442c4b6.png"> | <img width="300" src="https://user-images.githubusercontent.com/4588074/220423913-7ab5cc97-0215-4ab7-b96b-a8f573a0a657.png"> |

## Regression Notes
1. Potential unintended areas of impact
   Site creation → Domains when the experimental feature is disabled

3. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing

4. What automated tests I added (or what prevented me from doing so)
   Updated unit tests in `SiteCreationDomainsViewModelTest`. I don't think more tests are needed since most of the changes are in the view layer and the improved ui state strong typing is a solid way to guard against regressions when it comes to view-model.

PR submission checklist:
